keps/sig-autoscaling/0000-kep-template.md

```diff
@@ -0,0 +1,273 @@
+It's important to draw a line into the sand so only necessary portion of the cloud
+autoscaling logic is re-implemented instead of re-inventing entire wheel.
+
+### Cloud provider interface
+The cluster autoscaler defines two interfaces (`NodeGroup` and `CloudProvider`
+from `k8s.io/autoscaler/cluster-autoscaler/cloudprovider` package)
+that a cloud provider needs to implemented in order to allow the cluster
+autoscaler to properly perform scaling operations.
+
+The `CloudProvider` interface allows to:
+- operate on top of node groups
+- work with pricing models (to estimate reasonable expenses)
+- check limit resources (check if maximum resources per node group are exceeded)
+
+The `NodeGroup` interface allows to operate with node groups themselves:
+- change size of a node group
+- list nodes belonging to a node group
+- autoprovision new node group
+- generate node template for a node group (to determine resource capacity of
+  new node in case a node group has zero size)
+
+In case of `AWS` the `NodeGroup` then corresponds to autoscaling groups,
+in case of `Azure` to autoscaling sets, etc.
+
+### Node template
+
+In case a node group has no node, the cluster autoscaler needs a way to generate
+a node template for a given node group before it can predict how much new resources
+will get allocated and thus how many new nodes are needed to scale up.
+
+Each of the mentioned cloud providers has its own implementation of constructing
+the node template with its own list of information pulled from cloud provider API:
+
+* Azure:
+  - instance type to determine resource capacity (CPU, GPU, Memory)
+  - instance tags to construct node labels
+  - generic node labels: Arch, OS, instance type, region, failure domain, hostname
+* AWS:
+  - instance type to determine resource capacity (CPU, GPU, Memory)
+  - instance tags to construct node labels and taints
+  - generic node labels: Arch, OS, instance type, region, failure domain, hostname
+* GCE:
+  - instance type to determine resource capacity (CPU, Memory)
+  - instance tags to construct node labels (e.g. gpu.ResourceNvidiaGPU)
+  - maximum number of pods: apiv1.ResourcePods
+  - taints (from instance metadata)
+  - kubelet system and reserved allocatable
+
+## Code status
+
+At this point the cluster autoscaler does not support node autoprovisioning (TODO(jchaloup): link to the proposal).
+Every implementation assumes every node group already exists in the cloud provider.
+Both `Create` and `Delete` operations return an error `ErrAlreadyExist` and the autoprovisioning is off.
+
+The autoscaler allows to either enumerate all node groups (via `--nodes` option)
+or automatically discover all node groups on-the-fly (via `--node-group-auto-discovery`).
+
+## Integration with machine API
+
+Given the scaling up/down operation corresponds to increasing/decreasing the number
+of nodes in a node group, it's sufficient to "just" change the number of replicas
+of corresponding machine set.
```

Comment:
That is only true for scale-up. Scale-down requires the ability to delete a specific node, not just reduce number of replicas.

Reply:
Yes, enhancements for machine-set are being discussed at cluster-api community, which allows machine-set to delete specific machine while scaling-down.
- https://github.com/kubernetes-sigs/cluster-api/issues/75

Reply:
Yes, enhancements for machine-set are being discussed at cluster-api community, which allows machine-set to delete specific machine while scaling-down.

- kubernetes-sigs/cluster-api#75